s3/src/main/scala/akka/stream/alpakka/s3/impl/MemoryBuffer.scala

```diff
@@ -42,11 +43,14 @@ import akka.util.ByteString
       }
 
       override def onUpstreamFinish(): Unit = {
-        if (isAvailable(out)) emit()
+        if (isAvailable(out) && (!hasEmitted || buffer.nonEmpty)) emit()
         completeStage()
       }
 
-      private def emit(): Unit = emit(out, Chunk(Source.single(buffer), buffer.size), () => completeStage())
+      private def emit(): Unit = {
+        hasEmitted = true
+        emit(out, Chunk(Source.single(buffer), buffer.size), () => completeStage())
```

Comment:
An alternative impl. would be to empty the buffer as part of the  `andThen` thunk.  Then we don't need `hasEmitted` and just rely on the buffer to signal if there's anything to emit.